File: macos/Sources/dfdaemon/DiagnosticLog.swift
import Foundation

/// Opt-in diagnostic logging: off by default. Category-match counts are all
/// stats.json ever needed for the app to work, but debugging a specific
/// report ("X is wrongly blocked", "Y isn't blocked") needs to see actual
/// domain-level decisions — this exists only for that, only when the user
/// turns it on themselves, and never leaves the device automatically. See
/// docs/privacy.md.
final class DiagnosticLog {
    // Checking a file's mtime is cheap (a stat() call) — fine to do on every
    // query rather than caching, since toggling logging should take effect
    // immediately without a daemon restart.
    private static let maxBytes = 5 * 1024 * 1024 // rotate at 5MB — bounded, not unbounded growth
    private let lock = NSLock()
    private var lastEnabledCheck = Date.distantPast
    private var enabledCache = false
    private let checkInterval: TimeInterval = 2

    func log(_ line: String) {
        guard isEnabled() else { return }
        let timestamp = ISO8601DateFormatter().string(from: Date())
        let entry = "[\(timestamp)] \(line)\n"
        lock.lock()
        defer { lock.unlock() }
        rotateIfNeeded()
        if let data = entry.data(using: .utf8) {
            if let handle = FileHandle(forWritingAtPath: Config.diagnosticLogFile) {
                handle.seekToEndOfFile()
                handle.write(data)
                try? handle.close()
            } else {
                FileManager.default.createFile(atPath: Config.diagnosticLogFile, contents: data)
                try? FileManager.default.setAttributes([.posixPermissions: 0o644], ofItemAtPath: Config.diagnosticLogFile)
            }
        }
    }

    private func rotateIfNeeded() {
        guard let attrs = try? FileManager.default.attributesOfItem(atPath: Config.diagnosticLogFile),
              let size = attrs[.size] as? Int, size > Self.maxBytes else { return }
        try? FileManager.default.removeItem(atPath: Config.diagnosticLogFile)
    }

    private func isEnabled() -> Bool {
        let now = Date()
        lock.lock()
        if now.timeIntervalSince(lastEnabledCheck) < checkInterval {
            let cached = enabledCache
            lock.unlock()
            return cached
        }
        lock.unlock()

        let enabled = FileManager.default.fileExists(atPath: Config.loggingEnabledFile)
        lock.lock()
        enabledCache = enabled
        lastEnabledCheck = now
        lock.unlock()
        return enabled
    }
}

File: android/app/src/main/java/com/distractionfree/app/DiagnosticLog.kt
package com.distractionfree.app

import android.content.Context
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Opt-in diagnostic logging: off by default. Category-match counts are all
 * stats.json ever needed for the app to work, but debugging a specific
 * report ("X is wrongly blocked", "Y isn't blocked") needs to see actual
 * domain-level decisions — this exists only for that, only when the user
 * turns it on themselves, and never leaves the device automatically. See
 * docs/privacy.md. Mirrors the macOS daemon's DiagnosticLog.swift.
 */
class DiagnosticLog(private val context: Context) {
    companion object {
        private const val MAX_BYTES = 5 * 1024 * 1024 // rotate at 5MB
        private const val CHECK_INTERVAL_MS = 2000L
    }

    private val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
    @Volatile private var lastEnabledCheck = 0L
    @Volatile private var enabledCache = false
    private val writeLock = Object()

    fun log(line: String) {
        if (!isEnabled()) return
        val entry = "[${dateFormat.format(Date())}] $line\n"
        synchronized(writeLock) {
            try {
                val file = AppPaths.diagnosticLogFile(context)
                if (file.exists() && file.length() > MAX_BYTES) file.delete()
                file.appendText(entry)
            } catch (e: Exception) { /* best-effort, never crash the query path over logging */ }
        }
    }

    private fun isEnabled(): Boolean {
        val now = System.currentTimeMillis()
        if (now - lastEnabledCheck < CHECK_INTERVAL_MS) return enabledCache
        val enabled = AppPaths.loggingEnabledFile(context).exists()
        enabledCache = enabled
        lastEnabledCheck = now
        return enabled
    }
}
